runner/main.go

```diff
@@ -0,0 +1,269 @@
+		skipped   bool
+	}
+	internalTestResult struct {
+		ran        bool      // Test or benchmark (or one of its subtests) was executed.
+		failed     bool      // Test or benchmark has failed.
+		skipped    bool      // Test of benchmark has been skipped.
+		done       bool      // Test is finished and all subtests have completed.
+		finished   bool      // Test function has completed.
+		raceErrors int       // number of races detected during test
+		name       string    // Name of test or benchmark.
+		start      time.Time // Time test or benchmark started
+		duration   time.Duration
+	}
+)
+
+var runner *testRunner
+var runnerRegexName *regexp.Regexp
+
+// Runs a test suite
+func Run(m *testing.M, exitOnError bool, failRetriesCount int) int {
+	runner := &testRunner{
+		m:                m,
+		exitOnError:      exitOnError,
+		failRetriesCount: failRetriesCount,
+	}
+	runner.init()
+	return runner.Run()
+}
+
+// Gets the test name
+func GetTestName(name string) string {
+	if runnerRegexName == nil {
+		runnerRegexName = regexp.MustCompile(`(?m)([\w -:_]*)\/\[runner.[\w:]*](\/[\w -:_]*)?`)
+	}
+	match := runnerRegexName.FindStringSubmatch(name)
+	if match == nil || len(match) == 0 {
+		return name
+	}
+	return match[1] + match[2]
+}
+
+// Runs the test suite
+func (r *testRunner) Run() int {
+	r.exitCode = 0
+	r.m.Run()
+	return r.exitCode
+}
+
+// Internal test processor, each test calls the processor in order to handle retries, process exiting and exitcodes
+func (r *testRunner) testProcessor(t *testing.T) {
+	t.Helper()
+	name := t.Name()
+	if item, ok := (*r.tests)[name]; ok {
+
+		//Sets the original test name
+		if pointer, err := getFieldPointerOfT(t, "name"); err == nil {
+			*(*string)(pointer) = item.test.Name
+		}
+
+		run := 1
+		maxRetries := r.failRetriesCount
+		exitOnError := r.exitOnError
+		var rc interface{}
+
+		for {
+			var innerTest *testing.T
+			title := "Run"
+			if run > 1 {
+				title = "Retry:" + strconv.Itoa(run-1)
+			}
+			title = "[runner." + title + "]"
+			t.Run(title, func(it *testing.T) {
+				defer func() {
+					rc = recover()
+					if rc != nil {
+						it.FailNow()
+					}
+				}()
+				it.Helper()
+				innerTest = it
+				item.test.F(it)
+			})
+			innerTestInfo := r.getTestResultsInfo(innerTest)
+
+			if rc != nil {
+				if exitOnError {
+					panic(rc)
+				}
+				fmt.Println("PANIC RECOVER:", rc)
+				item.error = true
+			}
+			item.skipped = innerTestInfo.skipped
+			item.ran++
+
+			// Current run failure
+			if innerTestInfo.failed {
+				item.failed = true
+			}
+			// Current run ok but previous with fail -> Flaky
+			if !innerTestInfo.failed && item.failed {
+				item.failed = false
+				item.flaky = true
+				break
+			}
+
+			if item.skipped || !item.failed || run > maxRetries {
+				break
+			}
+			run++
+		}
+		if item.flaky {
+			fmt.Println("*** FLAKY", item.fqn)
+		}
+		if item.error && exitOnError {
+			panic(rc)
+		}
+		if item.error || item.failed {
+			r.exitCode = 1
```

Comment:
Why do we need to track this? Can we just rely on `m.Run`?

Reply:
Because `m.Run` return 1 if a sub test fail, so in flaky tests with both fail and pass status, always will be returning 1, and I think we want to have a green build besides flakyness. That`s why I have to control the exit code of the runner.

Reply:
forget it, I found a way to change the value in `m.Run` so we don't need to track the exitCode anymore.